pkg/stub/imagestreams.go

```diff
@@ -16,160 +17,90 @@ func (h *Handler) processImageStreamWatchEvent(is *imagev1.ImageStream, deleted
-		pending := false
-		anyErrors := false
-		// need to check for error conditions outside of the spec/status comparison because in an error scenario
-		// you can end up with less status tags than spec tags (especially if one tag refs another), but don't cite
-		// errors if we are now in the skip list; it is possible and imagestream is not in the skip list, we get an
-		// import error, and then a cluster admin puts the imagestream in the skip list
-		_, skipped := h.skippedImagestreams[is.Name]
-		if !skipped {
-			// so reaching this point means we have a prior upsert in progress, and we just want to track the status
-			logrus.Debugf("checking tag spec/status for %s spec len %d status len %d", is.Name, len(is.Spec.Tags), len(is.Status.Tags))
-
-			for _, statusTag := range is.Status.Tags {
-				// if an error occurred with the latest generation, let's give up as we are no longer "in progress"
-				// in that case as well, but mark the import failure
-				if statusTag.Conditions != nil && len(statusTag.Conditions) > 0 {
-					var latestGeneration int64
-					var mostRecentErrorGeneration int64
-					message := ""
-					for _, condition := range statusTag.Conditions {
-						if condition.Generation > latestGeneration {
-							latestGeneration = condition.Generation
-						}
-						if condition.Status == corev1.ConditionFalse &&
-							len(condition.Message) > 0 {
-							if condition.Generation > mostRecentErrorGeneration {
-								mostRecentErrorGeneration = condition.Generation
-								message = condition.Message
-							}
-						}
-					}
-					if mostRecentErrorGeneration >= latestGeneration {
-						logrus.Warningf("Image import for imagestream %s tag %s generation %v failed with detailed message %s", is.Name, statusTag.Tag, mostRecentErrorGeneration, message)
-						anyErrors = true
-						// add this imagestream to the Reason field
-						if !strings.Contains(importError.Reason, is.Name+" ") {
-							now := kapis.Now()
-							importError.Reason = importError.Reason + is.Name + " "
-							importError.Message = importError.Message + "<imagestream/" + is.Name + ">" + message + "<imagestream/" + is.Name + ">"
-							importError.Status = corev1.ConditionTrue
-							importError.LastTransitionTime = now
-							importError.LastUpdateTime = now
-							cfg.ConditionUpdate(importError)
-						}
-						break
+
+		// if we initiated the change, it will be in our cache, and we want to process
+		// the in progress, import error conditions
+		beforeErrorReason := cfg.Condition(v1.ImportImageErrorsExist).Reason
+		beforeProgressReason := cfg.Condition(v1.ImageChangesInProgress).Reason
+		beforeInReasons := strings.Contains(beforeProgressReason, is.Name+" ") ||
+			strings.Contains(beforeErrorReason, is.Name+"")
+		if cache.UpsertsAmount() > 0 {
+			cache.AddReceivedEventFromUpsert(is)
+			if !cfg.ConditionTrue(v1.ImageChangesInProgress) || !cfg.ConditionTrue(v1.SamplesExist) {
```

Comment:
should this be `if cfg.ConditionTrue(v1.ImageChangesInProgress) || !cfg.ConditionTrue(v1.SamplesExist)`  ?

(ie "if changesinprogress OR notsamplesexist")

Reply:
No.

in progress is not actually set to true until *AFTER* the samples are upserted ...so we can get events for those upserts before we set in progress to true, especially since cycling through upserting all the resources can take a bit

In case you don't recall, the "whether we set in progress to true before or after we start the actual samples upsert cycle" is a discussion we had with a few iterations earlier on.  Setting *AFTER* was deemed safest.

I generally go with !cfg.Condition... to count for the Unknown condition, which can occur when we have trouble communicating with the api server.

Reply:
thanks, yeah i can never keep that one in my head